src/stan/services/optimize/lbfgs.hpp

```diff
@@ -0,0 +1,204 @@
+#ifndef STAN_SERVICES_OPTIMIZE_LBFGS_HPP
+#define STAN_SERVICES_OPTIMIZE_LBFGS_HPP
+
+#include <stan/io/var_context.hpp>
+#include <stan/io/chained_var_context.hpp>
+#include <stan/io/random_var_context.hpp>
+#include <stan/callbacks/writer.hpp>
+#include <stan/model/util.hpp>
+#include <stan/optimization/bfgs.hpp>
+#include <stan/services/error_codes.hpp>
+#include <stan/services/util/initialize.hpp>
+#include <stan/services/util/rng.hpp>
+#include <fstream>
+#include <iostream>
+#include <iomanip>
+#include <string>
+#include <vector>
+
+
+namespace stan {
+  namespace services {
+    namespace optimize {
+
+      /**
+       * Runs the L-BFGS algorithm for a model.
+       *
+       * @tparam Model A model implementation
+       * @tparam Interrupt type of interrupt
+       *
+       * @param model Input model to test (with data already instantiated)
+       * @param init var context for initialization
+       * @param random_seed random seed for the pseudo random number generator
+       * @param chain chain id to advance the pseudo random number generator
+       * @param init_radius radius to initialize
+       * @param[in] history_size
+       * @param[in] init_alpha
+       * @param[in] tol_obj
+       * @param[in] tol_rel_obj
+       * @param[in] tol_grad
+       * @param[in] tol_rel_grad
+       * @param[in] tol_param
+       * @param[in] num_iterations maximum number of iterations
+       * @param[in] save_iterations indicates whether all the interations should
+       *   be saved to the parameter_writer
+       * @param[in] refresh how often to write output to message_writer
+       * @param[out] interrupt interrupt callback to be called every iteration
+       * @param[out] message_writer output for messages
+       * @param[out] init_writer Writer callback for unconstrained inits
+       * @param[out] parameter_writer output for parameter values
+       * @return stan::services::error_codes::OK (0) if successful
+       */
+      template <class Model, typename Interrupt>
+      int lbfgs(Model& model,
+                stan::io::var_context& init,
+                unsigned int random_seed,
+                unsigned int chain,
+                double init_radius,
+                int history_size,
+                double init_alpha,
+                double tol_obj,
+                double tol_rel_obj,
+                double tol_grad,
+                double tol_rel_grad,
+                double tol_param,
+                int num_iterations,
+                bool save_iterations,
+                int refresh,
+                Interrupt& interrupt,
```

Comment:
Yup. They're not a template parameter any more. They all have the same
signature.

On Fri, Aug 12, 2016 at 1:02 PM, Bob Carpenter <notifications@github.com>
wrote:

> In src/stan/services/optimize/lbfgs.hpp
> <https://github.com/stan-dev/stan/pull/2004#discussion_r74624797>:
>
> > +      int lbfgs(Model& model,
> > +                stan::io::var_context& init,
> > +                unsigned int random_seed,
> > +                unsigned int chain,
> > +                double init_radius,
> > +                int history_size,
> > +                double init_alpha,
> > +                double tol_obj,
> > +                double tol_rel_obj,
> > +                double tol_grad,
> > +                double tol_rel_grad,
> > +                double tol_param,
> > +                int num_iterations,
> > +                bool save_iterations,
> > +                int refresh,
> > +                Interrupt& interrupt,
>
> Is Interrupt a typedef or a template parameter? If the interrupts are only
> to interrupt, then I don't see why they couldn't have the same signature
> for all routes and be regular arguments rather than template parameters. -
> Bob
> … <#m_-8438171675509282684_>
> On Aug 12, 2016, at 4:49 PM, Krzysztof Sakrejda ***@***.***> wrote: In
> src/stan/services/optimize/lbfgs.hpp: > + int lbfgs(Model& model, > +
> stan::io::var_context& init, > + unsigned int random_seed, > + unsigned int
> chain, > + double init_radius, > + int history_size, > + double init_alpha,
> > + double tol_obj, > + double tol_rel_obj, > + double tol_grad, > + double
> tol_rel_grad, > + double tol_param, > + int num_iterations, > + bool
> save_iterations, > + int refresh, > + Interrupt& interrupt, @syclik
> <https://github.com/syclik> The type of the interrupt callback here is
> different from the interrupt callback type for
> sample/fixed_param_sampler.hpp (or similar file name). I started writing
> tests for the sample/* services functions but I can't tell if these are
> supposed to be the same or not. — You are receiving this because you were
> mentioned. Reply to this email directly, view it on GitHub, or mute the
> thread.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/stan/pull/2004/files/feda4a2510d7528741b04badba3ec0cf629733f4#r74624797>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAZ_F1-mWznoqgbahO7ILi6ti1emq7ahks5qfKcSgaJpZM4Jfpda>
> .
>